### PR TITLE
add ability to specify api key

### LINKED ISSUE #720: Implement API Key authentication
From https://github.com/elastic/apm/issues/183

Also make sure to update the docs and the setup instructions in Kibana.

### PR DESCRIPTION
**Pull request by vhatsura**
resolves #720 

TBD:
- [x] documentation
- [x] tests
- [ ] test it with apm server

### CODE REVIEW

**Comment by vhatsura** [comment]
@bmorelli25, could you check the documentation, please? I used documentation to go apm agent as a reference.

@gregkalapos, do we need validation for the `ApiKey` option? Does it have proper format and valid base64 encoding? I don't see such validations in other agents

**Comment by lreuven** [comment]
Very cool,
Thanks Vadim.
we started adding cucumber tests for new features ( have a look [here](https://github.com/elastic/apm-agent-ruby/pull/655/commits/33d07922419e73010df4fe48028f40ae3be2291b)
any chance you can align the tests to the same specs?
and once again, thanks for the amazing work ❤️

**Comment by gregkalapos** [comment]
➕ Thanks @vhatsura for the PR! 

> do we need validation for the ApiKey option? Does it have a proper format and valid base64 encoding? I don't see such validations in other agents

I don't see it neither in other agents, nor in the cucumber specs. Overall I don't think this is something agents necessarily need to align on. If we don't validate, then the HTTP Request to the server will fail - if we validate then we could print a more specific log message about the config.

We could try to convert it one time (make sure it happens only once) `byte[] HashBytes = Convert.FromBase64String(Value)` if that fails it's invalid. If it fails it could revert to `SecretToken` , if that is not set, then... well then either just don't send anything to the server or send it, which will fail.

I personally feel aligning on these cases is not super important, but feel free to jump in on https://github.com/elastic/apm/issues/183 and ask other agent devs.


> In addition, I believe it would be better to rethink the configuration architecture inside agent because it is needed to update a few IConfigurationReader implementations for introducing one option

Yes, this already came up https://github.com/elastic/apm-agent-dotnet/issues/626#issuecomment-558476964 - I agree on that, currently we technically break all implementation when we add a new config, which is not ideal, this could be improved.

**Comment by vhatsura** [comment]
@lreuven 

> we started adding cucumber tests for new features ( have a look here
any chance you can align the tests to the same specs?

I saw it but didn't know about alignments. Will try to look into it

**Comment by bmorelli25** [comment]
@elasticmachine, run elasticsearch-ci/docs

File: docs/configuration.asciidoc [hunk 1]
```diff
@@ -393,6 +393,35 @@ This string is used to ensure that only your agents can send data to your APM se
 Both the agents and the APM server have to be configured with the same secret token.
 Use this setting in case the APM Server requires a token (e.g. APM Server in Elastic Cloud).
 
+WARNING: the API Key is sent as plain-text in every request to the server, so you should also secure
```
Comment (bmorelli25) [inline docs/configuration.asciidoc]:
```suggestion
WARNING: the `SecretToken` is sent as plain-text in every request to the server, so you should also secure
```

File: docs/configuration.asciidoc [hunk 2]
```diff
@@ -393,6 +393,35 @@ This string is used to ensure that only your agents can send data to your APM se
...
+|============
+
+This base64-encoded string is used to ensure that only your agents can send data to your APM server.
+You must have created the API key using the APM server command line tool. Please see the APM server
+{apm-app-ref}/api-key.html#create-api-key[create-api-key] documentation for details on how to do that.
+
+
+NOTE: This feature is fully supported in the APM Server versions >= 7.6.
+
+WARNING: the API Key is sent as plain-text in every request to the server, so you should also secure
```
Comment (bmorelli25) [inline docs/configuration.asciidoc]:
```suggestion
WARNING: the `APIKey` is sent as plain-text in every request to the server, so you should also secure
```

File: docs/configuration.asciidoc [hunk 3]
```diff
@@ -393,6 +393,35 @@ This string is used to ensure that only your agents can send data to your APM se
...
+|============
+
+[options="header"]
+|============
+| Default                 | Type
+| `<none>`                | A base64-encoded string
+|============
+
+This base64-encoded string is used to ensure that only your agents can send data to your APM server.
+You must have created the API key using the APM server command line tool. Please see the APM server
```
Comment (bmorelli25) [inline docs/configuration.asciidoc]:
```suggestion
You must have created the API key using the APM server's {apm-server-ref-v}/api-key.html[command line tool].
```

File: docs/configuration.asciidoc [hunk 4]
```diff
@@ -393,6 +393,35 @@ This string is used to ensure that only your agents can send data to your APM se
...
+{apm-app-ref}/api-key.html#create-api-key[create-api-key] documentation for details on how to do that.
```
Comment (bmorelli25) [inline docs/configuration.asciidoc]:
```suggestion
```

**Comment by bmorelli25** [comment]
@elasticmachine, run elasticsearch-ci/docs

**Comment by vhatsura** [comment]
@lreuven, tests were added

**Comment by bmorelli25** [comment]
@elasticmachine, run elasticsearch-ci/docs

File: test/Elastic.Apm.Feature.Tests/Features/api_key.feature [hunk 5]
```diff
@@ -0,0 +1,18 @@
+Feature: Api Key
```
Comment (gregkalapos) [inline test/Elastic.Apm.Feature.Tests/Features/api_key.feature]: Ideally long term it'd be nice if these could be shared across agents. But for now we don't really have a solution for that.

**Comment by vhatsura** [comment]
> I went through this, thanks @vhatsura!
> 
> Looks ok to me.
> 
> Regarding "test it with apm server" - that'll be done outside this repo, we already opened
> [elastic/apm-integration-testing#781](https://github.com/elastic/apm-integration-testing/issues/781). I don't see that as a blocker for this PR, I think we are good to merge this already.

Sorry, totally forgot about excluding auto-generated files. Did it and added to git ignore list

**Comment by gregkalapos** [comment]
Unfortunately we have CI fails here on `net461`:

```
[2020-03-04T00:46:13.715Z] [xUnit.net 00:00:09.93]     A configured api key is sent in the Authorization header [FAIL]
[2020-03-04T00:46:13.715Z] [xUnit.net 00:00:09.93]       System.IO.FileLoadException : Could not load file or assembly 'Elastic.Apm, Version=1.3.1.0, Culture=neutral, PublicKeyToken=ae7400d2c189cf22' or one of its dependencies. Strong name signature could not be verified.  The assembly may have been tampered with, or it was delay signed but not fully signed with the correct private key. (Exception from HRESULT: 0x80131045)
[2020-03-04T00:46:13.715Z] [xUnit.net 00:00:09.94]       Stack Trace:
[2020-03-04T00:46:13.715Z] [xUnit.net 00:00:09.94]            at Elastic.Apm.Feature.Tests.FeatureContexts.ApiKeyFeatureContext.<CheckAuthorizationHeader>d__5.MoveNext()
[2020-03-04T00:46:13.715Z] [xUnit.net 00:00:09.94]            at System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[TStateMachine](TStateMachine& stateMachine)
[2020-03-04T00:46:13.715Z] [xUnit.net 00:00:09.94]            at Elastic.Apm.Feature.Tests.FeatureContexts.ApiKeyFeatureContext.CheckAuthorizationHeader(String authorizationHeader)
[2020-03-04T00:46:13.715Z] [xUnit.net 00:00:09.94]         ApiKeySteps.cs(34,0): at Elastic.Apm.Feature.Tests.ApiKeySteps.<ThenAuthorizationHeaderIs>d__9.MoveNext()
[2020-03-04T00:46:13.715Z] [xUnit.net 00:00:09.94]         --- End of stack trace from previous location where exception was thrown ---
[2020-03-04T00:46:13.715Z] [xUnit.net 00:00:09.94]            at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
[2020-03-04T00:46:13.715Z] [xUnit.net 00:00:09.94]            at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
[2020-03-04T00:46:13.715Z] [xUnit.net 00:00:09.94]            at TechTalk.SpecFlow.Bindings.SynchronousBindingDelegateInvoker.<>c__DisplayClass2_0.<<InvokeBindingDelegateAsync>b__0>d.MoveNext()
 ```

Even worse, I was unable to repro the issue locally on a Windows machine.

Probably some strong naming issue, maybe `StrongNamer` would solve this, but overall I think for now it's fine if to only target `netcoreapp2.1` in `Elastic.Apm.Feature.Tests`.

I'll simply remove the `net461` target for that test and merge - happy to follow up, but I don't feel this should hold us back.

**Comment by gregkalapos** [comment]
Something is still strange here. I saw some failing GC metrics tests, which are unrelated - those are fixed in another PR.

But then when the GC test does not fail the CI still runs into timeout after 1 hour, because the .NET Core Windows tests ~45 minutes and then CI reaches the 1 hour limit and kills the test - I suspect something does not work on Windows, but I don't see yet what it is.

**Comment by gregkalapos** [comment]
Merging now, failing test should be unrelated.

File: test/Elastic.Apm.Feature.Tests/Features/api_key.feature (hunk 5 above)
Comment (SylvainJuge) [reply test/Elastic.Apm.Feature.Tests/Features/api_key.feature]: There is a shell script that just does this within python agent: https://raw.githubusercontent.com/beniwohli/apm-agent-python/a393b5f0e522ccbeef9a5d79e1efcc181f2180cc/tests/scripts/download_gherkin_features.sh

I plan do use something similar for Java agent implementation

Comment (mdelapenya) [reply test/Elastic.Apm.Feature.Tests/Features/api_key.feature]: For the record, we are in the process of automating this for all the agents :)

**Pull request merged**